src/plugins/es_ui_shared/static/forms/hook_form_lib/types.ts

```diff
@@ -24,21 +24,37 @@ import { Subject, Subscription } from './lib';
 // Comes from https://github.com/microsoft/TypeScript/issues/15012#issuecomment-365453623
 type Required<T> = T extends FormData ? { [P in keyof T]-?: NonNullable<T[P]> } : T;
 
-export interface FormHook<T extends FormData = FormData> {
+export interface FormHook<T extends FormData = FormData, I extends FormData = T> {
+  /**  Flag that indicates if the form has been submitted at least once. It is set to `true` when we call `submit()`. */
   readonly isSubmitted: boolean;
+  /** Flag that indicates if the form is being submitted. */
   readonly isSubmitting: boolean;
+  /** Flag that indicates if the form is valid. It can have three values: `true | false | undefined`. */
```

Comment:
nit; I would phrase this as:

```ts
/** Flag that indicates if the form is valid. If `undefined` then form validation has not been checked yet. */
```

I think the TS `boolean` value is self-explanatory for the other states.